### PR TITLE
Explain editing of additional app charts

### LINKED ISSUE #1271: Document Editing of Helm chart(s)
**Is your feature request related to a problem? Please describe.**
For the new application helm charts, we should provide some mechanism to push updates or edit in place

**Describe the solution you'd like**
I'd ideally like to see a way to edit charts similar to `kubectl edit` or an editor in the UI

**Describe alternatives you've considered**
We could give access to the s3 bucket and let operators edit there but it seems dangerous

**Additional context**
This doesn't necessarily need to be done in the epinio cli but that's likely the easiest place to do this.

### PR DESCRIPTION
**Pull request by andreas-kupries**
Fix [#1271](https://github.com/epinio/epinio/issues/1271).

### CODE REVIEW

File: docs/howtos/create_custom_appcharts.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+```
+$ cd ../..
+$ helm package chart/application
+```
+
+The tarball is placed into the current working directory and has the form
+`application-VERSION.tar.gz` where `VERSION` is the chart version.
+
+Note that in this How-To we have not attempted to change the chart version. Versioned
+chart development is out of scope here.
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:95]:
```suggestion
> Note: we didn't change the chart version. Versioned
chart development is out of scope for this How-To.
```

File: docs/howtos/create_custom_appcharts.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+
+## Making the helm Chart known to Epinio
+
+The first step in making the new chart created in the previous section available to Epinio
+is to place the generated tarball on some publicly accessible server.
+
+Given the plethora of possible options this How-To simply assumes that the tarball is
+available at the example URL
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:104]:
```suggestion
Given the plethora of possible options, this How-To simply assumes that the tarball is
available at the example URL:
```

File: docs/howtos/create_custom_appcharts.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+```
+https://mydomain.org/epinio-application-0.1.15-fluentd.tgz
+```
+
+Now save
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:110]:
```suggestion
Copy the following text, change the `RELEASE_NAMESPACE` value to the namespace where Epinio was
installed into (by default it's `epinio`) and add it to a file of your choice:
```

File: docs/howtos/create_custom_appcharts.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  namespace: RELEASE_NAMESPACE
+  name: fluentd
+spec:
+  shortDescription: Fluentd filterable standard deployment
+  description: Epinio standard support chart extended for fluent filtering
+  helmChart: https://mydomain.org/epinio-application-0.1.15-fluentd.tgz
+```
+
+into a file of your choice and change `RELEASE_NAMESPACE` to be the namespace Epinio was
+installed into. The default is `epinio`.
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:125]:
```suggestion
```

File: docs/howtos/create_custom_appcharts.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+$ kubectl apply -f /path/to/your/file
+```
+
+Verify that the new chart is now known to Epinio by invoking
+
+```
+$ epinio app chart list
+```
+
+and/or
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:144]:
```suggestion
You can also see the details of the chart in Epinio:
```

File: docs/howtos/create_custom_appcharts.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+```
+$ epinio app chart show fluent
+```
+
+If this fails check that the `RELEASE_NAMESPACE` was properly set in the CR.
```
Comment (nunix) [inline docs/howtos/create_custom_appcharts.md:150]:
```suggestion
## Troubleshooting
If the new Helm chart is not correctly registered in Epinio, check that the `RELEASE_NAMESPACE` was properly set in the CRD.
```

**Review by nunix — COMMENTED** [review]
Reviewed `create_custom_appcharts` only. Will have a second review for the `using_custom_appcharts`.

File: docs/howtos/create_custom_appcharts.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+$ cd helm-charts/chart/application
+```
+
+In this How-To we now create a variant of the chart which adds an annotation to every
+application `Deployment` which enables the filtering of Epinio applications in
+[fluentd](https://www.fluentd.org/).
+
+Note that explanations about setting up and using `fluentd` are out of scope here.
+
+Use your editor of choice to edit the file `templates/deployment.yaml`.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply docs/howtos/create_custom_appcharts.md:43]: In the clone step one of the commands has the user setting `helm-charts/chart/application` as their working directory.
Thus the shorter relative path here is ok.

We could possibly rewrite the entire how-to to stay either outside the clone, or on the toplevel directory, and then use larger paths across the board. On the good side of that change the user would not have to track state in the form of the working directory.

File: docs/howtos/create_custom_appcharts.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        fluentd-application-name: {{ .Values.epinio.appName }}
+        app.kubernetes.io/name: {{ .Values.epinio.appName }}
+[...]
+```
+
+Note how the templating uses the `.Values.epinio.appName` field to insert the application
+name into the annotation.
+
+The file `values.yaml` in the chart explains the full set of values Epinio sets when
+deploying an application through the chart.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply docs/howtos/create_custom_appcharts.md:81]: See previous response regarding working directory in the executed commands.

File: docs/howtos/create_custom_appcharts.md (hunk 6 above)
Comment (andreas-kupries) [reply docs/howtos/create_custom_appcharts.md:150]: `CRD` is the custom resource definition. `CR` is the custom resource and what we have here. Maybe better to expand the shorthand, or name it differently.

**Comment by andreas-kupries** [comment]
My plan is now to accept all these, and then re-tweak a few places I noted.

File: docs/howtos/using_custom_appcharts.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+# Querying the default Helm chart
+
+To query if there is a currently active default, and what it is set to invoke the command
+
+```
+$ epinio app chart default
+```
+
+Note that the configured default is also made visible when listing the available charts.
+In the example output below the chart `foo` is marked as the currently active default.
```
Comment (nunix) [inline docs/howtos/using_custom_appcharts.md:90]:
```suggestion
You can also check which chart is set as `default` when listing the available charts.
As an example, the output below shows the chart `FOO` as the current default:
```

File: docs/howtos/using_custom_appcharts.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+```
+| DEFAULT |   NAME   |        DESCRIPTION         |
+|---------|----------|----------------------------|
+| *       | foo      | Foofy deployment           |
```
Comment (nunix) [inline docs/howtos/using_custom_appcharts.md:95]:
```suggestion
| *       | FOO      | Foofy deployment           |
```

File: docs/howtos/using_custom_appcharts.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+|         | standard | Epinio standard deployment |
+```
+
+
+# Unsetting the default Helm chart
+
+To restore the use of the system default (`standard`) invoke the command
```
Comment (nunix) [inline docs/howtos/using_custom_appcharts.md:102]:
```suggestion
You can set the system chart `standard` back as the default, by running the following command:
```

File: docs/howtos/using_custom_appcharts.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+```
+$ epinio app chart default ''
+```
+
+Note the __empty string__ used in place of the chart name. This is the signal to unset the
+current setting.
```
Comment (nunix) [inline docs/howtos/using_custom_appcharts.md:109]:
```suggestion
:::note
The __empty string__ used in place of the chart name is mandatory. If you do not add it,
the command will only list the charts available.
:::
```

**Review by nunix — COMMENTED** [review]
Reviewed the `using_custom_appcharts` doc.
Rephrased to active voice and added admonitions.

File: docs/howtos/using_custom_appcharts.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+  - [epinio app create](../references/cli/epinio_app_create.md)
+  - [epinio app update](../references/cli/epinio_app_update.md)
+  - [epinio push](../references/cli/epinio_push.md)
+
+The argument of the option is the name of custom Helm chart to use, as shown when listing
+the available charts.
+
+__Beware__ that changing the chart to use is __not possible__ if the application has an
+active workload.
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply docs/howtos/using_custom_appcharts.md:50]: Thanks, was not aware of this kind of syntax.

File: docs/howtos/using_custom_appcharts.md [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+$ epinio app chart list
+```
+
+For an unmodified Epinio installation the output will show only the standard chart Epinio
+comes with:
+
+```
+| DEFAULT |   NAME   |        DESCRIPTION         |
+|---------|----------|----------------------------|
+|         | standard | Epinio standard deployment |
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply docs/howtos/using_custom_appcharts.md:31]: No. `standard` is the __system default__ and is automatically used when the user neither sets a local default, nor uses `--app-chart`. As such it is not set as the local default by default, and is not highlighted.

**Comment by andreas-kupries** [comment]
Thanks for the review. I accept all the changes and will then fix a few typos I saw in them.

**Comment by andreas-kupries** [comment]
Pushed my fixes to your fixes now.

**Review by nunix — APPROVED** [review]

**Pull request merged**